### PR TITLE
experimental option to override transformer dtype during inference

### PR DESCRIPTION
**Pull request by logogin**
First functional change to override transformer dtype. Marked cli option as experimental.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [ ] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [ ] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [ ] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/arguments.py [hunk 1]
```diff
@@ -1067,6 +1067,11 @@ def add_inference_args(params):
                                type=float,
                                help='Beta factor for the length penalty used in beam search: '
                                     '(beta + len(Y))**alpha/(beta + 1)**alpha. Default: %(default)s')
+    decode_params.add_argument('--override-dtype',
+                               default=None,
+                               type=str,
+                               help='EXPERIMENTAL: may be changed or removed in future. Forces dtype of encoders and '
```
Comment (fhieber) [inline sockeye/arguments.py]: 'Forces' -> 'Enforces' or 'Overrides training dtype of encoders and decoders during inference.'

File: sockeye/arguments.py [hunk 2]
```diff
@@ -1067,6 +1067,11 @@ def add_inference_args(params):
...
+                                    'decoders during inference. Default: %(default)s')
```
Comment (fhieber) [inline sockeye/arguments.py]: [supernit] Period missing :)

File: sockeye/encoder.py [hunk 3]
```diff
@@ -932,6 +933,7 @@ def encode(self,
         :param seq_len: Maximum sequence length.
         :return: Encoded versions of input data data, data_length, seq_len.
         """
+        data = mx.sym.cast(data=data, dtype=self.dtype)
```
Comment (fhieber) [inline sockeye/encoder.py]: as said in my previous comment, this code is also used in training and without the override-dtype flag, so we should make sure that (no-op) casting does not incur any performance penalties.

File: sockeye/transformer.py [hunk 4]
```diff
@@ -291,14 +293,15 @@ def __init__(self, max_length: int) -> None:
     def forward(self, is_train, req, in_data, out_data, aux):
         # lengths: (batch_size,)
         lengths = in_data[0]
+        dtype = lengths.dtype
 
         # (batch_size, max_length)
-        data = mx.nd.zeros((lengths.shape[0], self.max_length), ctx=lengths.context)
+        data = mx.nd.zeros((lengths.shape[0], self.max_length), dtype=dtype, ctx=lengths.context)
         data = mx.nd.SequenceMask(data=data,
```
Comment (fhieber) [inline sockeye/transformer.py]: the bias tensor is added to the data in the encoder. Are you sure we are not causing any fp16 overflow AFTER this addition?

File: sockeye/transformer.py [hunk 5]
```diff
@@ -291,14 +293,15 @@ def __init__(self, max_length: int) -> None:
     def forward(self, is_train, req, in_data, out_data, aux):
```
Comment (fhieber) [inline sockeye/transformer.py]: transformer.py also contains `AutoRegressiveBias` which also uses a large negative value to ensure auto-regressiveness in the Transformer Decoder. This also needs to be fixed.

File: sockeye/inference.py [hunk 6]
```diff
@@ -371,6 +372,8 @@ def load_models(context: mx.context.Context,
                                         vocabulary.  Used when logits/softmax are handled separately.
     :param cache_output_layer_w_b: Models cache weights and biases for logit computation as NumPy arrays (used with
                                    restrict lexicon).
+    :param override_dtype: Overrides dtype of encoder and decoder to a different from one, defined in config
```
Comment (fhieber) [inline sockeye/inference.py]: Maybe "Overrides dtype of encoder and decoder defined at training time to a different one."?

File: sockeye/transformer.py [hunk 7]
```diff
@@ -69,6 +69,7 @@ class TransformerEncoderBlock:
     def __init__(self,
                  config: TransformerConfig,
                  prefix: str) -> None:
+        self.dtype = config.dtype
```
Comment (fhieber) [inline sockeye/transformer.py]: why is this needed? couldn't we use the (potentially overriden) `self.config.dtype`?

File: sockeye/transformer.py [hunk 8]
```diff
@@ -120,6 +121,7 @@ class TransformerDecoderBlock:
...
```
Comment (fhieber) [inline sockeye/transformer.py]: same here

File: sockeye/transformer.py (hunk 7 above)
Comment (logogin) [reply sockeye/transformer.py]: you are right, missed that config is a member already.

File: sockeye/encoder.py (hunk 3 above)
Comment (logogin) [reply sockeye/encoder.py]: I will try to double-ckeck that. Will be something to report as a bug in MxNet if no-op casting will have penalties.

File: sockeye/arguments.py (hunk 2 above)
Comment (logogin) [reply sockeye/arguments.py]: where? After "experimental"?

File: sockeye/transformer.py (hunk 4 above)
Comment (logogin) [reply sockeye/transformer.py]: actually, this is just a fix for the custom operator. Due to the zeros symbol without dtype it was forced always to be fp32. But in the change, bias calculated in fp32 and casted to fp16 before addition.

File: sockeye/arguments.py (hunk 2 above)
Comment (fhieber) [reply sockeye/arguments.py]: 'Default: %(default)s' -> 'Default: %(default)s.' :)

File: sockeye/transformer.py (hunk 4 above)
Comment (fhieber) [reply sockeye/transformer.py]: I am not sure I get what you are saying. I understand that the custom op now creates data in the correct dtype. The problem is that you are masking this fp16 data in some places with the smallest possible value in the fp16 range. Now when you actually use this bias tensor in the transformer model, it gets added to the encoder representation tensors. Depending on how they look like, this addition might cause an overflow: `fp16.min + -x`. 
So instead of using a value for masking that is right at the limit of the representable numeric range, it might be better to use something less extreme.

File: sockeye/transformer.py (hunk 5 above)
Comment (fhieber) [reply sockeye/transformer.py]: Its the same argument as in my comment above. you need to guard against hitting the fp16 range at masking time, and when using the bias tensor in the model.

Comment (logogin) [reply sockeye/transformer.py]: added unit test and support for dtype argument

File: sockeye/transformer.py (hunk 4 above)
Comment (logogin) [reply sockeye/transformer.py]: Changed `float16` constant to be something close to maximum, but not maximum possible number.

File: sockeye/transformer.py [hunk 9]
```diff
@@ -399,9 +401,10 @@ def backward(self, req, out_grad, in_data, out_data, in_grad, aux):
 @mx.operator.register("auto_regressive_bias")
 class AutoRegressiveBiasProp(mx.operator.CustomOpProp):
 
-    def __init__(self, length: str) -> None:
+    def __init__(self, length: str, dtype: Optional[str] = C.DTYPE_FP32) -> None:
```
Comment (fhieber) [inline sockeye/transformer.py]: I don't think `dtype` could be None, so `Optional` does not apply here.

Comment (logogin) [reply sockeye/transformer.py]: fixed
